packages/core/src/query/flux.ts

```diff
@@ -0,0 +1,245 @@
+  return new FluxParameter(`duration(v: "${sanitizeString(value)}")`)
+}
+
+function sanitizeRegExp(value: any): string {
+  return `regexp.compile(v: "${sanitizeString(value)}")`
+}
+
+/**
+ * Creates flux regexp literal.
+ */
+export function fluxRegExp(value: any): FluxParameterLike {
+  // let the server decide if it can be parsed
+  return new FluxParameter(sanitizeRegExp(value))
+}
+
+/**
+ * Creates flux boolean literal.
+ */
+export function fluxBool(value: any): FluxParameterLike {
+  if (value === 'true' || value === 'false') {
+    return new FluxParameter(value)
+  }
+  return new FluxParameter((!!value).toString())
+}
+
+/**
+ * Assumes that the supplied string is a flux literal value and creates
+ * a parameter out of it.
+ */
+export function fluxLiteral(value: string): FluxParameterLike {
+  // let the server decide if it can be parsed
+  return new FluxParameter(value)
+}
+
+/**
+ * Escapes content of the supplied parameter so that it can be safely embedded into flux query.
+ * @param value parameter value
+ * @return sanitized flux value or an empty string if it cannot be converted
+ */
+export function toFluxValue(value: any): string {
+  if (value === undefined) {
+    return ''
+  } else if (value === null) {
+    return 'null'
+  } else if (typeof value === 'boolean') {
+    return value.toString()
+  } else if (typeof value === 'string') {
+    return `"${sanitizeString(value)}"`
+  } else if (typeof value === 'number') {
+    return sanitizeFloat(value)
+  } else if (
+    typeof value === 'object' &&
+    typeof value[FLUX_VALUE] === 'function'
+  ) {
+    return value[FLUX_VALUE]()
+  } else if (value instanceof Date) {
+    return value.toISOString()
+  } else if (value instanceof RegExp) {
+    return sanitizeRegExp(value)
+  }
+  return toFluxValue(value.toString())
+}
+
+/**
+ * Flux is a tagged template that sanitizes supplied parameters
+ * to avoid injection attacks in flux.
+ */
+export function flux(
+  strings: TemplateStringsArray,
+  ...values: any
+): ParameterizedQuery {
+  if (strings.length == 1) return strings[0] // the simplest case
+  const parts = new Array<string>(strings.length + values.length)
+  let partIndex = 0
+  for (let i = 0; i < strings.length; i++) {
+    const text = strings[i]
+    parts[partIndex++] = text
+    if (i < values.length) {
+      const val = values[i]
+      let sanitized: string
+      if (
+        text.endsWith('"') &&
+        i + 1 < strings.length &&
+        strings[i + 1].startsWith('"')
+      ) {
+        // parameter is wrapped into flux double quotes
+        sanitized = sanitizeString(val)
+      } else {
+        sanitized = toFluxValue(val)
+        if (sanitized === '') {
+          throw new Error(
+            `Unsupported parameter literal '${val}' at index: ${i}, type: ${typeof val}`
+          )
+        }
+      }
+      parts[partIndex++] = sanitized
+    } else if (i < strings.length - 1) {
```

Comment:
why is this checked on every loop and not once at the start?
`values.length < strings.length - 1`?

Reply:
it is not checked on every loop, it is checked at most once and only if required to do so